File: KASIKORN Business-Technology Group/Assignment1/Ex2/Ex2.py
import random
from datetime import datetime, timedelta
import csv
import time
import psutil  # Import psutil for memory monitoring
import os  # Import os for file path handling

class Ex2:
    def __init__(self):
        self.current_date = datetime.now()
        self.corporate_branches = ['80001']
        self.individual_branches = ['90001']
        self.provinces = ['HN', 'HCM', 'DN', 'HP', 'CT']
        self.positions = ['Manager', 'Staff', 'Engineer', 'Accountant', 'Director']
        self.company_sizes = ['S', 'M', 'L']
    
    def generate_date_of_birth(self):
        age = random.randint(20, 60)
        birth_date = self.current_date - timedelta(days=age*365 + random.randint(0, 365))
        return birth_date.strftime('%Y-%m-%d')
    
    def generate_individual_customer(self):
        dob = self.generate_date_of_birth()
        customer_id = f"I{random.randint(1000000, 9999999)}"
        
        return {
            'TransactionDate': self.current_date.strftime('%Y-%m-%d'),
            'BRID': random.choice(self.individual_branches),
            'CUSTID': customer_id,
            'BANKCODE': '79601001',
            'RESIDENT': random.choice(['0', '1']),
            'CUST_TYPE': 'P02',  # Individual
            'BIRTHDAY': dob,
            'FULL_NAME': f"Individual Customer {customer_id}",
            'PHONE': f"0{random.randint(100000000, 999999999)}",
            'EMAIL': f"customer_{customer_id}@email.com",
            'POSITION': random.choice(self.positions),
            'INCOME': random.randint(5000000, 50000000),
            'PROVINCE_CODE': random.choice(self.provinces),
            'COUNTRY_CODE': 'VN',
            'IDENTITY_NUM': f"{random.randint(100000000, 999999999)}",
            'TOTAL_ASSETS': random.randint(100000000, 1000000000),
            'TOTAL_INCOME': random.randint(50000000, 500000000),
            'TOTAL_LIABILITY': random.randint(10000000, 100000000)
        }

    def generate_corporate_customer(self):
        customer_id = f"C{random.randint(1000000, 9999999)}"
        
        return {
            'TransactionDate': self.current_date.strftime('%Y-%m-%d'),
            'BRID': random.choice(self.corporate_branches),
            'CUSTID': customer_id,
            'BANKCODE': '79601001',
            'RESIDENT': '1',
            'CUST_TYPE': 'P01',  # Corporate
            'FULL_NAME': f"Corporation {customer_id}",
            'PHONE': f"0{random.randint(100000000, 999999999)}",
            'EMAIL': f"corp_{customer_id.lower()}@corp.com",
            'PROVINCE_CODE': random.choice(self.provinces),
            'COUNTRY_CODE': 'VN',
            'TOTAL_ASSETS': random.randint(1000000000, 10000000000),
            'TOTAL_INCOME': random.randint(500000000, 5000000000),
            'TOTAL_LIABILITY': random.randint(100000000, 1000000000),
            'BIRTHDAY': ''
        }

    def generate_file(self, num_records, filename):
        start_time = time.time()
        start_memory = psutil.virtual_memory().used / (1024 * 1024)  # Memory in MB
        
        num_corporate = int(num_records * 0.05)         
        num_individual = num_records - num_corporate 
        
        fieldnames = list(self.generate_individual_customer().keys())
        
        with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames, delimiter=',')
            writer.writeheader()
            
            for _ in range(num_corporate):
                writer.writerow(self.generate_corporate_customer())
                
            for _ in range(num_individual):
                writer.writerow(self.generate_individual_customer())
        
        end_time = time.time()
        end_memory = psutil.virtual_memory().used / (1024 * 1024)  # Memory in MB
        
        return end_time - start_time, end_memory - start_memory

def main():
    generator = Ex2()
    record_counts = [10000, 100000, 1000000]
    
    print("Starting generate test data...")
    print(f"{'Records':<10} {'Time (seconds)':<15} {'Speed (records/sec)':<20} {'Memory Used (MB)'}")
    print("-" * 100)
    
    for count in record_counts:
        filename = f'Customer_extract_{datetime.now().strftime("%Y%m%d")}_{count}.csv'
        execution_time, memory_used = generator.generate_file(count, filename)
        speed = count / execution_time
        print(f"{count:<10} {execution_time:<15.2f} {speed:<20.2f} {memory_used:.2f}")
    
    print("-" * 100)

if __name__ == "__main__":
    main()

File: KASIKORN Business-Technology Group/Assignment1/Ex2/Ex2.java
package Assignment1.Ex2;

import java.io.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class Ex2 {
    private final LocalDateTime currentDate;
    private final List<String> corporateBranches;
    private final List<String> individualBranches;
    private final List<String> provinces;
    private final List<String> positions;
    private final List<String> companySizes;
    private final Random random;

    public Ex2() {
        this.currentDate = LocalDateTime.now();
        this.corporateBranches = Collections.singletonList("80001");
        this.individualBranches = Collections.singletonList("90001");
        this.provinces = Arrays.asList("HN", "HCM", "DN", "HP", "CT");
        this.positions = Arrays.asList("Manager", "Staff", "Engineer", "Accountant", "Director");
        this.companySizes = Arrays.asList("S", "M", "L");
        this.random = new Random();
    }

    private String generateDateOfBirth() {
        int age = random.nextInt(41) + 20; // 20 to 60
        LocalDateTime birthDate = currentDate.minusDays(age * 365L + random.nextInt(365));
        return birthDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
    }

    private String generateRandomDate(int minDays, int maxDays) {
        LocalDateTime date = currentDate.minusDays(random.nextInt(maxDays - minDays) + minDays);
        return date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
    }

    private Map<String, String> generateCustomer(boolean isCorporate) {
        String customerId = (isCorporate ? "C" : "I") + String.format("%07d", random.nextInt(9000000) + 1000000);
        String incDate = generateRandomDate(365, 7300);

        Map<String, String> customer = new LinkedHashMap<>();
        customer.put("TransactionDate", currentDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")));
        customer.put("BRID", isCorporate ? corporateBranches.get(0) : individualBranches.get(0));
        customer.put("CUSTID", customerId);
        customer.put("BANKCODE", "79601001");
        customer.put("RESIDENT", isCorporate ? "1" : (random.nextBoolean() ? "1" : "0"));
        customer.put("CUST_TYPE", isCorporate ? "P01" : "P02");
        customer.put("FULL_NAME", (isCorporate ? "Corporation " : "Individual Customer ") + customerId);
        customer.put("ADDRESS", isCorporate ? "Corporate Address" : "Personal Address");
        customer.put("PHONE", String.format("0%09d", random.nextInt(900000000) + 100000000));
        customer.put("EMAIL", String.format("%s_%s@email.com", isCorporate ? "corp" : "customer", customerId.toLowerCase()));
        customer.put("PROVINCE_CODE", provinces.get(random.nextInt(provinces.size())));
        customer.put("COUNTRY_CODE", "VN");
        customer.put("IDENTITY_NUM", String.format("%09d", random.nextInt(900000000) + 100000000));
        customer.put("IDENTITY_DATE", isCorporate ? incDate : generateRandomDate(365, 3650));
        customer.put("SIZEOFCO", isCorporate ? companySizes.get(random.nextInt(companySizes.size())) : "");
        customer.put("TOTAL_ASSETS", String.valueOf(random.nextInt(900000001) + 100000000));
        customer.put("TOTAL_INCOME", String.valueOf(random.nextInt(450000001) + 50000000));
        customer.put("TOTAL_LIABILITY", String.valueOf(random.nextInt(90000001) + 10000000));
        return customer;
    }

    public double generateFile(int numRecords, String filename) throws IOException {
        long startTime = System.currentTimeMillis();
        Runtime runtime = Runtime.getRuntime();
        long initialMemory = runtime.totalMemory() - runtime.freeMemory();

        int numCorporate = (int) (numRecords * 0.05);
        int numIndividual = numRecords - numCorporate;

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            writer.write(String.join(",", generateCustomer(false).keySet()));
            writer.newLine();

            for (int i = 0; i < numCorporate; i++) {
                writer.write(String.join(",", generateCustomer(true).values()));
                writer.newLine();
            }
            for (int i = 0; i < numIndividual; i++) {
                writer.write(String.join(",", generateCustomer(false).values()));
                writer.newLine();
            }
        }

        long finalMemory = runtime.totalMemory() - runtime.freeMemory();
        double executionTime = (System.currentTimeMillis() - startTime) / 1000.0;
        double memoryUsage = (finalMemory - initialMemory) / (1024.0 * 1024.0);
        System.out.printf("Memory used: %.2f MB\n", memoryUsage);
        return executionTime;
    }

    public static void main(String[] args) {
        Ex2 generator = new Ex2();
        int[] recordCounts = {10000, 100000, 1000000};

        System.out.println("Starting generate test data...");
        System.out.printf("%-10s %-15s %-15s %-15s%n", "Records", "Time (sec)", "Speed (rec/sec)", "Memory (MB)");
        System.out.println("-".repeat(80));

        DateTimeFormatter fileFormatter = DateTimeFormatter.ofPattern("yyyyMMdd");
        String dateStr = LocalDateTime.now().format(fileFormatter);
        Runtime runtime = Runtime.getRuntime();

        for (int count : recordCounts) {
            String filename = String.format("Customer_extract_%s_%d.csv", dateStr, count);
            try {
                long initialMemory = runtime.totalMemory() - runtime.freeMemory();
                double executionTime = generator.generateFile(count, filename);
                double speed = count / executionTime;
                long finalMemory = runtime.totalMemory() - runtime.freeMemory();
                double memoryUsage = (finalMemory - initialMemory) / (1024.0 * 1024.0);
                System.out.printf("%-10d %-15.2f %-15.2f %-15.2f%n", count, executionTime, speed, memoryUsage);
            } catch (IOException e) {
                System.err.println("Error generating file: " + e.getMessage());
            }
        }
        System.out.println("-".repeat(80));
    }
}
